### PR TITLE
chore: initial merge conflict resolution for update up to f5aa05e

### PR DESCRIPTION
**Pull request by elfedy**
Initial merge conflic resolution for upstream update to [f5aa05e](https://github.com/foundry-rs/foundry/commit/f5aa05ee9ac4f088db64b93ae5df4d8656faf1ea)
Compilation errors are not solved yet.

### Updated forked dependencies:
*  `alloy` to `0.4.2`: https://github.com/Karrq/alloy/tree/zksync-v0.4.2
* `compilers` to `0.11.4`: https://github.com/Moonsong-Labs/compilers/tree/zksync-v0.11.4
* `foundry-fork-db` to `0.4.0`: https://github.com/Moonsong-Labs/foundry-zksync-fork-db/tree/zksync-v0.4.0 (Added a fix due to a deprecated method)
* `block-explorers` still at `0.7.3` but points to new `compilers`

### Conflicting files:

- .github/workflows/test.yml 
- Cargo.lock
- Cargo.toml 
- Dockerfile 
- README.md 
- crates/cheatcodes/src/evm.rs 
- crates/cheatcodes/src/evm/mock.rs 
- crates/cheatcodes/src/inspector.rs 
- crates/cheatcodes/src/test.rs 
- crates/cheatcodes/src/test/expect.rs 
- crates/evm/abi/src/console/hardhat.rs 
- crates/evm/core/src/backend/cow.rs 
- crates/evm/core/src/backend/mod.rs 
- crates/evm/core/src/decode.rs 
- crates/evm/core/src/lib.rs 
- crates/evm/core/src/utils.rs 
- crates/evm/evm/src/executors/fuzz/mod.rs 
- crates/evm/evm/src/executors/mod.rs 
- crates/evm/fuzz/src/strategies/param.rs 
- crates/evm/traces/src/decoder/mod.rs 
- crates/evm/traces/src/identifier/local.rs 
- crates/forge/Cargo.toml 
- crates/forge/src/result.rs 
- crates/forge/tests/cli/script.rs 
- crates/script/Cargo.toml 
- crates/script/src/broadcast.rs 
- crates/script/src/lib.rs 
- crates/script/src/sequence.rs 
- crates/script/src/simulate.rs 
- crates/script/src/transaction.rs 
- testdata/cheats/Vm.sol 

### Conflicting commits
New cheatcodes:

https://github.com/foundry-rs/foundry/pull/8571: record traces
https://github.com/foundry-rs/foundry/pull/9024: mockCalls
https://github.com/foundry-rs/foundry/pull/8952: gasSnapshots over arbitrary sections
https://github.com/foundry-rs/foundry/pull/9087: cheatcodes to derive and save multiple wallets
https://github.com/foundry-rs/foundry/pull/9106: mv script wallets into Cheatcode
https://github.com/foundry-rs/foundry/pull/8883: Warnings for deprecated cheatcodes:

Fixes:

https://github.com/foundry-rs/foundry/pull/9050: Identify artifacts using both deployed and creation code.
https://github.com/foundry-rs/foundry/pull/9096: running script with `--broadcast` for a transaction sequence can error out due to nonce desync from rpc latency 

Refactors:

https://github.com/foundry-rs/foundry/pull/8924 dyn DatabaseExt vs generic refactor
https://github.com/foundry-rs/foundry/pull/9011 Reduce dynamic dispatch from inspectors
https://github.com/foundry-rs/foundry/pull/8969 Remove unreachable decoding of expect revert
https://github.com/foundry-rs/foundry/pull/9098 move `ScriptSequence` to new crate
https://github.com/foundry-rs/foundry/pull/8945 renaming of snapshot to be more specific

CI

https://github.com/foundry-rs/foundry/pull/8948: Add makefile and codespell
https://github.com/foundry-rs/foundry/pull/8964: use multiarchitecture images in Dockerfile
https://github.com/foundry-rs/foundry/commit/7a9ebf9ccbce2957762ef1b3f4623efb76ef0306: Bump alpine to 3.20

Deps

Bump alloy core to 0.8.26: https://github.com/foundry-rs/foundry/pull/9045

### CODE REVIEW

**Comment by tomimor** [comment]
For more visibility on the update process @dutterbutter @popzxc

File: crates/evm/evm/src/inspectors/stack.rs [hunk 1]
```diff
@@ -889,64 +983,49 @@ impl<'a, DB: DatabaseExt> Inspector<DB> for InspectorStackRefMut<'a> {
...
-            self,
-            ecx
         );
 
         false
     }
 
     fn console_log(&mut self, input: String) {
-        call_inspectors!([&mut self.log_collector], |inspector| InspectorExt::<DB>::console_log(
+        call_inspectors!([&mut self.log_collector], |inspector| InspectorExt::console_log(
```
Comment (nbaztec) [inline crates/evm/evm/src/inspectors/stack.rs:1028]: Same needs to be applied to `fn trace_zksync` below

File: crates/script/src/lib.rs [hunk 2]
```diff
@@ -574,12 +572,16 @@ impl ScriptConfig {
...
+<<<<<<< HEAD
         cheats_data: Option<(
             ContractsByArtifact,
             ScriptWallets,
             ArtifactId,
             DualCompiledContracts,
         )>,
+=======
+        cheats_data: Option<(ContractsByArtifact, Wallets, ArtifactId)>,
+>>>>>>> foundry/master
```
Comment (nbaztec) [inline crates/script/src/lib.rs:584]: Pending merge

```rs
cheats_data: Option<(ContractsByArtifact, Wallets, ArtifactId, DualCompiledContracts)>,
```

File: crates/script/src/lib.rs [hunk 3]
```diff
@@ -635,6 +638,24 @@ impl ScriptConfig {
...
+        if let Some((known_contracts, script_wallets, target)) = cheats_data {
+            builder = builder.inspectors(|stack| {
+                stack
+                    .cheatcodes(
+                        CheatsConfig::new(
+                            &self.config,
+                            self.evm_opts.clone(),
+                            Some(known_contracts),
+                            Some(target.name),
+                            Some(target.version),
```
Comment (nbaztec) [inline crates/script/src/lib.rs:651]: Missing

```rs
dual_compiled_contracts,
use_zk,
```

File: crates/script/src/lib.rs [hunk 4]
```diff
@@ -635,6 +638,24 @@ impl ScriptConfig {
...
+                        )
+                        .into(),
+                    )
+                    .wallets(script_wallets)
+                    .enable_isolation(self.evm_opts.isolate)
+            });
```
Comment (nbaztec) [inline crates/script/src/lib.rs:657]: Pending merge

```rs
.use_zk_vm(use_zk);
```

**Review by nbaztec — APPROVED** [review]

**Pull request merged**